### PR TITLE
DAOS-11479 control: Add dmg system exclude

### PR DESCRIPTION
**Pull request by mjmac**
Adds new command surface to allow an admin to move
a rank into or out of the AdminExcluded state:
  * dmg system exclude --ranks=<ranklist>
  * dmg system clear-exclude --ranks=<ranklist>

When a rank has been placed in this state, it cannot
join the system or be used for new pools.

Also cleans up some test code to avoid leaking goroutines,
as they make debugging test failures more difficult than
necessary. Fixes a bug in the dRPC failure logic that
caused the local raft instance to shut down instead of
just resigning leadership, as this could leave the system
in a bad state and require a restart of daos_server.

Required-githooks: true

Signed-off-by: Michael MacDonald <mjmac.macdonald@intel.com>

### CODE REVIEW

**Comment by mjmac** [comment]
This isn't quite ready yet. Will get it working correctly on the CR branch and then port it to master.

**Review by kjacque — APPROVED** [review]
LGTM. The suggestion is very minor.

File: src/control/server/instance_exec.go [hunk 1]
```diff
@@ -226,7 +213,24 @@ func (ei *EngineInstance) Run(ctx context.Context, recreateSBs bool) {
...
+				if runnerExitCh != nil {
+					restartRequested = true
+					continue
+				}
+
+				runnerExitCh, err = ei.startRunner(ctx, recreateSBs)
+				if err != nil {
+					ei.log.Errorf("runner exited without starting process: %s", err)
+					ei.handleExit(ctx, 0, err)
+					continue
```
Comment (tanabarr) [inline src/control/server/instance_exec.go:225]: nit: unnecessary continue statement

**Review by tanabarr — APPROVED** [review]
Changes look good.

The engine runner refactor doesn't seem to be directly related to adding the system exclude commands so it should be mentioned in the commit message as it touches on significant internal operations.

Was the aforementioned refactor a reaction to a bug or just a general improvement?

**Comment by mjmac** [comment]
> Was the aforementioned refactor a reaction to a bug or just a general improvement?

This patch was extracted from work done on the catastrophic recovery branch in order to support restarting ranks in a special checker mode if they haven't been admin excluded. What I found with the previous code is that I couldn't reliably restart ranks if they'd exited with an error during the join process (they would stay stuck in the Errored state and require a restart of daos_server to get them fixed). There were a number of weird corner cases and after playing whack-a-mole for a while, I decided it would be easier to break things down a bit to more clearly define areas of responsibility.

I'll update the commit message to include some more detail about that part of the work.

**Pull request merged**